File: src/podcast/services/feed.py
from feedgen.feed import FeedGenerator
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from podcast.config import settings
from podcast.models import Episode, PodcastSettings


def _format_duration(seconds: int) -> str:
    """Format seconds as HH:MM:SS for iTunes duration."""
    hours = seconds // 3600
    minutes = (seconds % 3600) // 60
    secs = seconds % 60
    if hours > 0:
        return f"{hours}:{minutes:02d}:{secs:02d}"
    return f"{minutes}:{secs:02d}"


async def generate_feed(db: AsyncSession) -> str:
    """Generate a podcast-compatible RSS feed."""
    podcast_settings = await db.get(PodcastSettings, 1)
    if not podcast_settings:
        podcast_settings = PodcastSettings()

    base = settings.base_url.rstrip("/")

    fg = FeedGenerator()
    fg.load_extension("podcast")

    # Channel-level metadata
    fg.title(podcast_settings.title)
    fg.description(podcast_settings.description)
    fg.link(href=f"{base}/feed.xml", rel="self")
    fg.link(href=base, rel="alternate")
    fg.language(podcast_settings.language)
    fg.generator("Podcast Generator")

    # iTunes metadata
    fg.podcast.itunes_author(podcast_settings.author)
    fg.podcast.itunes_summary(podcast_settings.description)
    fg.podcast.itunes_category("Technology")
    fg.podcast.itunes_explicit("no")

    image_url = podcast_settings.image_url or f"{base}/static/til.png"
    fg.podcast.itunes_image(image_url)

    # Episodes
    result = await db.execute(
        select(Episode)
        .where(Episode.status == "ready")
        .where(Episode.audio_filename.isnot(None))
        .order_by(Episode.published_at.desc())
    )
    episodes = result.scalars().all()

    for ep in episodes:
        fe = fg.add_entry()
        fe.id(str(ep.id))
        fe.title(ep.title)
        fe.description(ep.description or ep.topic)
        fe.published(ep.published_at)

        audio_url = f"{base}/audio/{ep.audio_filename}"
        fe.enclosure(audio_url, str(ep.audio_size_bytes or 0), "audio/mpeg")

        if ep.audio_duration_seconds:
            fe.podcast.itunes_duration(_format_duration(ep.audio_duration_seconds))

        if ep.episode_number:
            fe.podcast.itunes_episode(ep.episode_number)

        fe.podcast.itunes_author(podcast_settings.author)

    return fg.rss_str(pretty=True).decode("utf-8")

File: src/services/feed.ts
import { Feed } from "feed";
import { eq, and, isNotNull } from "drizzle-orm";
import { db } from "../database.js";
import { episodes, podcastSettings, type User } from "../schema.js";
import { settings } from "../config.js";

function formatDuration(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);
  const secs = seconds % 60;
  if (hours > 0) {
    return `${hours}:${String(minutes).padStart(2, "0")}:${String(secs).padStart(2, "0")}`;
  }
  return `${minutes}:${String(secs).padStart(2, "0")}`;
}

export async function generateFeed(user: User): Promise<string> {
  const [psettings] = await db
    .select()
    .from(podcastSettings)
    .where(eq(podcastSettings.userId, user.id))
    .limit(1);

  const ps = psettings ?? {
    title: "My Private Podcast",
    description: "AI-generated podcast episodes",
    author: "Podcast Bot",
    language: "en",
    imageUrl: null,
  };

  const base = settings.baseUrl.replace(/\/$/, "");
  const imageUrl = ps.imageUrl || `${base}/static/til.png`;

  const feed = new Feed({
    title: ps.title,
    description: ps.description,
    id: `${base}/feed/${user.feedToken}.xml`,
    link: base,
    language: ps.language,
    generator: "Podcast Generator",
    image: imageUrl,
    copyright: "",
    author: { name: ps.author },
  });

  // Fetch ready episodes
  const readyEpisodes = await db
    .select()
    .from(episodes)
    .where(
      and(
        eq(episodes.userId, user.id),
        eq(episodes.status, "ready"),
        isNotNull(episodes.audioFilename)
      )
    )
    .orderBy(episodes.publishedAt);

  for (const ep of readyEpisodes) {
    const audioUrl = `${base}/audio/${user.id}/${ep.audioFilename}`;
    feed.addItem({
      title: ep.title,
      id: ep.id,
      link: `${base}/episodes/${ep.id}`,
      description: ep.description || ep.topic,
      date: ep.publishedAt ?? ep.createdAt,
      enclosure: {
        url: audioUrl,
        length: ep.audioSizeBytes ?? 0,
        type: "audio/mpeg",
      },
      extensions: [
        {
          name: "_itunes",
          objects: {
            author: ps.author,
            ...(ep.audioDurationSeconds
              ? { duration: formatDuration(ep.audioDurationSeconds) }
              : {}),
            ...(ep.episodeNumber ? { episode: String(ep.episodeNumber) } : {}),
          },
        },
      ],
    });
  }

  return feed.rss2();
}
